modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DataNodesManager.java

```diff
@@ -0,0 +1,1527 @@
+     * {@link DistributionZonesUtil#zoneScaleUpTimerKey(int)} and {@link DistributionZonesUtil#zoneScaleDownTimerKey}
+     * if it passes the condition. It is called on the first creation of a zone.
+     *
+     * @param zoneId Unique id of a zone
+     * @param timestamp Timestamp of an event that has triggered this method.
+     * @param dataNodes Data nodes.
+     * @return Future reflecting the completion of initialisation of zone's keys in meta storage.
+     */
+    CompletableFuture<Void> onZoneCreate(
+            int zoneId,
+            HybridTimestamp timestamp,
+            Set<NodeWithAttributes> dataNodes
+    ) {
+        if (!busyLock.enterBusy()) {
+            throw new IgniteInternalException(NODE_STOPPING_ERR, new NodeStoppingException());
+        }
+
+        try {
+            // Update data nodes for a zone only if the corresponding data nodes keys weren't initialised in ms yet.
+            Condition condition = and(
+                    notExists(zoneDataNodesHistoryKey(zoneId)),
+                    notTombstone(zoneDataNodesHistoryKey(zoneId))
+            );
+
+            Update update = ops(
+                    addNewEntryToDataNodesHistory(zoneId, new DataNodesHistory(), timestamp, dataNodes),
+                    clearTimer(zoneScaleUpTimerKey(zoneId)),
+                    clearTimer(zoneScaleDownTimerKey(zoneId)),
+                    clearTimer(zonePartitionResetTimerKey(zoneId))
+            ).yield(true);
+
+            Iif iif = iif(condition, update, ops().yield(false));
+
+            return metaStorageManager.invoke(iif)
+                    .thenApply(StatementResult::getAsBoolean)
+                    .whenComplete((invokeResult, e) -> {
+                        if (e != null) {
+                            LOG.error(
+                                    "Failed to initialize zone's dataNodes history [zoneId = {}, timestamp = {}, dataNodes = {}]",
+                                    e,
+                                    zoneId,
+                                    timestamp,
+                                    nodeNames(dataNodes)
+                            );
+                        } else if (invokeResult) {
+                            LOG.info("Initialized zone's dataNodes history [zoneId = {}, timestamp = {}, dataNodes = {}]",
+                                    zoneId,
+                                    timestamp,
+                                    nodeNames(dataNodes)
+                            );
+                        } else {
+                            LOG.debug(
+                                    "Failed to initialize zone's dataNodes history [zoneId = {}, timestamp = {}, dataNodes = {}]",
+                                    zoneId,
+                                    timestamp,
+                                    nodeNames(dataNodes)
+                            );
+                        }
+                    }).thenCompose((ignored) -> nullCompletedFuture());
```

Comment:
Why is this line needed?

Reply:
1. `.thenCompose(v -> nullCompletedFuture()` looks very strange, `thenApply(v -> null)` makes more sense to me in this context.
2. You can simply change the return type from `CompletableFuture<Void>` to `CompletableFuture<?>` and remove this line. I don't like that we need to do something in runtime just to make the types coerce.